File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for(int i = 0 ; i < nums.length && nums[i] <= 0 ; i++) {
            if(i > 0 && nums[i] == nums[i-1]) continue;
            this.twoSum(i,nums,res);
            
        }
        
        return res;
    }
    private void twoSum(int index,int[] nums,List<List<Integer>> res) {
        int left = index + 1, right = nums.length - 1;
        
        while(left < right) {
            long sum = nums[left] + nums[right] + nums[index];
            if(sum < 0) {
                left++;
            } else if (sum > 0) {
                right--;
            } else {
                res.add(Arrays.asList(nums[index],nums[left++],nums[right--]));
                //What happen if the left number is repeated?
                //It will add the same list for each repetition. So, we need to avoid that
                while(left < right && nums[left] == nums[left - 1]) left++;
            }
        }
        
    }
}

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    
    nums.sort((a,b) => a-b);
    let res = [];
    let twoSum = (i) => {
        let lo = i + 1;
        let hi = nums.length - 1;
        
        while(lo < hi) {
            let sum = nums[i] + nums[lo] + nums[hi];
            
            if(sum < 0) {
                lo++;
            } else if(sum>0) {
                hi--;
            } else {
                res.push([nums[i] ,nums[lo++] ,nums[hi--]]);
                while(lo<hi && nums[lo] === nums[lo - 1])
                    lo++;
            }
        }
    }
    
    
    for(let i = 0; i < nums.length && nums[i] <= 0; i++) {
        if(i === 0 || nums[i - 1] !== nums [i]) {
            twoSum(i);
        }
    }
    return res;
};